examples/algorithms/sycl/include/sycl/track_finding/seeding_algorithm.hpp

```diff
@@ -0,0 +1,72 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include <chrono>
+#include <iostream>
+
+#include "../../../device/sycl/include/sycl/seeding/seed_finding.hpp"
+#include "../../../device/sycl/include/sycl/seeding/spacepoint_binning.hpp"
+
+namespace traccc {
+namespace sycl {
+
+class seeding_algorithm
+    : public algorithm<host_seed_container(host_spacepoint_container&&)> {
+
+    public:
+    seeding_algorithm(vecmem::memory_resource& mr, ::sycl::queue* q = nullptr)
+        : m_mr(mr), m_q(q) {
+
+        m_config.highland = 13.6 * std::sqrt(m_config.radLengthPerSeed) *
+                            (1 + 0.038 * std::log(m_config.radLengthPerSeed));
+        float maxScatteringAngle = m_config.highland / m_config.minPt;
+        m_config.maxScatteringAngle2 = maxScatteringAngle * maxScatteringAngle;
+        // helix radius in homogeneous magnetic field. Units are Kilotesla, MeV
+        // and millimeter
+        // TODO: change using ACTS units
+        m_config.pTPerHelixRadius = 300. * m_config.bFieldInZ;
+        m_config.minHelixDiameter2 =
+            std::pow(m_config.minPt * 2 / m_config.pTPerHelixRadius, 2);
+        m_config.pT2perRadius =
+            std::pow(m_config.highland / m_config.pTPerHelixRadius, 2);
+
+        m_grid_config.bFieldInZ = m_config.bFieldInZ;
+        m_grid_config.minPt = m_config.minPt;
+        m_grid_config.rMax = m_config.rMax;
+        m_grid_config.zMax = m_config.zMax;
+        m_grid_config.zMin = m_config.zMin;
+        m_grid_config.deltaRMax = m_config.deltaRMax;
+        m_grid_config.cotThetaMax = m_config.cotThetaMax;
+
+        sb = std::make_shared<traccc::sycl::spacepoint_binning>(
+            traccc::sycl::spacepoint_binning(m_config, m_grid_config, mr, q));
+        sf = std::make_shared<traccc::sycl::seed_finding>(
+            traccc::sycl::seed_finding(m_config, sb->nbins(), mr, q));
+    }
+
+    output_type operator()(
+        host_spacepoint_container&& spacepoints) const override {
+        output_type seeds({host_seed_container(1, &m_mr.get())});
+        auto internal_sp_g2 = sb->operator()(std::move(spacepoints));
+        seeds =
+            sf->operator()(std::move(spacepoints), std::move(internal_sp_g2));
```

Comment:
? You `std::move` the `spacepoints` variable twice. This is **very** undefined behaviour. :frowning: And it nicely shows the pitfalls of being way too enthusiastic with moving objects around... (Khm... @stephenswat... khm...) Please resolve this.

And at the same time you should simplify the incantations here a bit.

```c++
auto internal_sp_g2 = sb(spacepoints);
seeds = sf(spacepoints, internal_sp_g2);
```

should probably work, no?

Reply:
So here, both operators in `seed_finding` and `spacepoint_binning` accept an R-value reference, so that's why there is `std::move` applied on them, and both the `sb` and `sf` are shared pointers to the `spacepoint_binning` and `seed_finding` functors, so directly calling `sb(spacepoints` and `sf(spacepoints, internal_sp_g2)` does not really work on them - this was taken directly from the CUDA code :stuck_out_tongue:

Reply:
Both of those functors/algorithms are yours. If the CUDA code defined them with rvalue references, that should be a problem for that code. But in the SYCL version let's just ingest const lvalue references in those algorithms instead. It will make the memory management of the code a whole lot simpler.

For the smart pointers: Yeah, I forgot about that. Though I'm also not entirely sure why we need to use smart pointers to instantiate these functors, and can't just do it by value? :confused:

In any case, as long as these are smart pointer variables, you could write:

```c++
auto internal_sp_g2 = (*sb)(spacepoints);
seeds = (*sf)(spacepoints, internal_sp_g2);
```

Reply:
I'm having some troubles :stuck_out_tongue: I changed passing the `spacepoints` to the `const &` but then inside the `seed_finding.hpp`, inside  the functions that have the kernel calls inside them (such as [doublet_counting](https://github.com/konradkusiak97/traccc/blob/9f005a70144b3024337336b811e227ab7b05d8e4/device/sycl/src/seeding/doublet_counting.sycl)), I'm doing:
```cpp
void doublet_counting(const seedfinder_config& config, const sp_grid& internal_sp,
                      host_doublet_counter_container& doublet_counter_container,
                      vecmem::memory_resource& resource, ::sycl::queue* q) {
...
    auto internal_sp_view = get_data(internal_sp, resource);
```
And it doesn't allow me to call `get_data()` on the `const sp_grid&` type
```shell
[ 94%] Building SYCL object device/sycl/CMakeFiles/traccc_sycl.dir/src/seeding/populating_grid.sycl.o
/home/kkusiak/traccc-test/traccc/device/sycl/src/seeding/doublet_counting.sycl:150:29: error: no matching function for call to 'get_data'
    auto internal_sp_view = get_data(internal_sp, resource);
                            ^~~~~~~~
/home/kkusiak/traccc-test/traccc/core/include/edm/container.hpp:369:41: note: candidate template ignored: could not match 'container' against 'grid2'
inline container_data<header_t, item_t> get_data(
                                        ^
/home/kkusiak/traccc-test/build_nvidia/_deps/detray-src/core/include/detray/grids/grid2.hpp:477:1: note: candidate template ignored: deduced type 'grid2<...>' of 1st parameter does not match adjusted type 'const grid2<...>' of argument [with populator_type = detray::attach_populator, axis_p0_type = detray::axis::circular, axis_p1_type = detray::axis::regular, serializer_type = detray::serializer2, vector_type = detray::dvector, jagged_vector_type = detray::djagged_vector, array_type = detray::darray, tuple_type = detray::dtuple, value_type = traccc::internal_spacepoint<traccc::spacepoint>, kSORT = false, kDIM = 1]
get_data(grid2<populator_type, axis_p0_type, axis_p1_type, serializer_type,
^
/home/kkusiak/traccc-test/build_nvidia/_deps/detray-src/core/include/detray/grids/axis.hpp:656:54: note: candidate function template not viable: requires single argument 'axis', but 2 arguments were provided
inline axis_data<axis_type<array_type, vector_type>> get_data(
```
which is defined in [traccc/core/include/edm/container.hpp](https://github.com/konradkusiak97/traccc/blob/9f005a70144b3024337336b811e227ab7b05d8e4/core/include/edm/container.hpp)
I have the recent changes in the latest [commit](https://github.com/acts-project/traccc/pull/120/commits/9f005a70144b3024337336b811e227ab7b05d8e4)

Reply:
Hmm... Unfortunately this highlights a shortcoming in the current code. :frowning: It comes from a pretty nuanced issue in how vector views work in [vecmem](https://github.com/acts-project/vecmem).

The "problem" is sort of summarised in this code:

https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/containers/vector.hpp#L30-L37

```c++
/// Helper function creating a @c vecmem::data::vector_view object
template <typename TYPE, typename ALLOC>
VECMEM_HOST data::vector_view<TYPE> get_data(std::vector<TYPE, ALLOC>& vec);

/// Helper function creating a @c vecmem::data::vector_view object
template <typename TYPE, typename ALLOC>
VECMEM_HOST data::vector_view<const TYPE> get_data(
    const std::vector<TYPE, ALLOC>& vec);
```

You see that we defined 2 `get_data(...)` functions for `vecmem::vector`. The first one is simple, it translates a non-const `vecmem::vector<T>` object into a `vecmem::data::vector_view<T>` one. But for a `const vecmem::vector<T>` object we can't just make a `const vecmem::data::vector_view<T>` object. Since we pass the view objects around mostly by value, that const declaration would be lost very quickly.

For this reason a constant vector is represented by a `vecmem::data::vector_view<const T>` data type. Hardcoding in the type that we are not allowed to modify the vector's elements through this view.

Once you switched to passing the spacepoint grid as a const reference to the function, the currently defined `get_data(...)` function (coming [from detray](https://github.com/acts-project/detray/blob/main/core/include/detray/grids/grid2.hpp#L457-L482)) became unable to return a proper view type. :frowning: @beomki-yeo and @niermann999, we will need to sort this out in the detray code.

For now let's just do the easy/dirty thing in this PR. Remove `const` from the lvalue reference, and just give a non-const reference to this function. It's a pretty dirty setup for the code, but let's be pragmatic. Once the "infrastructure code" becomes smarter, we'll be able to make the code more const-correct.

Reply:
Hm, but can I actually do that? :thinking: Because the `seed_finding` and `spacepoint_binning` algorithms are inheriting from the `algorithm` class defined in [traccc/core/include/utils/algorithm.hpp](https://github.com/acts-project/traccc/blob/main/core/include/utils/algorithm.hpp) where it ensures that the `rvalue reference` or the ` const lvalue reference` is being used. I'm getting failures of `static_assert` when I'm trying to just use the `lvalue reference` :confused:

Reply:
Then let's keep the interface with the const reference, and do another type of ugly hack. `const_cast` away the constness of the reference inside of that function.

A safer option would be to make a copy of the object inside of this function. But the code is slow enough as is...

Reply:
> Though I'm also not entirely sure why we need to use smart pointers to instantiate these functors

 When I tried to use value, it required copy assignment operator for functors. For example;
` sb = traccc::spacepoint_binning(m_config, m_grid_config, mr)); // need copy assignment operator`
 
 and yeah i was too lazy to add that

> where it ensures that the rvalue reference or the const lvalue reference is being used

Yeah, that's why I used rvalue in CUDA code. It seems not so desirable pattern according to @krasznaa 's explanation

Reply:
That's also done.